vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh

```diff
@@ -91,4 +101,63 @@ disableSystemdTimesyncdAndEnableNTP() {
 
     # enable ntp
     systemctlEnableAndStart ntp || exit $ERR_NTP_START_TIMEOUT
-}
\ No newline at end of file
+}
+
+installFIPS() {
+    echo "Installing FIPS..."
+    wait_for_apt_locks
+
+    echo "adding ua repository..."
+    retrycmd_if_failure 5 10 120 add-apt-repository -y ppa:ua-client/stable || exit $ERR_ADD_UA_APT_REPO
+    apt_get_update || exit $ERR_APT_UPDATE_TIMEOUT
+
+    echo "installing ua tools..."
+    apt_get_install 5 10 120 ubuntu-advantage-tools || exit $ERR_UA_TOOLS_INSTALL_TIMEOUT
+
+    echo "auto attaching ua..."
+    retrycmd_if_failure 5 10 120 ua auto-attach || exit $ERR_AUTO_UA_ATTACH
+
+    echo "disabling ua livepatch..."
+    retrycmd_if_failure 5 10 300 echo y | ua disable livepatch
+
+    echo "enabling ua fips..."
+    retrycmd_if_failure 5 10 1200 echo y | ua enable fips || exit $ERR_UA_ENABLE_FIPS
+
+    echo "installing strongswan..."
+    apt_get_install 5 10 120 strongswan=5.6.2-1ubuntu2.fips.2.4.2 || exit $ERR_STRONGSWAN_INSTALL_TIMEOUT
+
+    # workaround to make GPU provisioning in CSE work
+    # under /usr/src/linux-headers-4.15.0-1002-azure-fips there are some dangling symlinks to non-existing linux-azure-headers-4.15.0-1002
+    # this causes command '/usr/sbin/dkms build -m nvidia -v 450.51.06 -k 4.15.0-1002-azure-fips' for GPU provisioning in CSE to fail
+    # however linux-headers-4.15.0-1002-azure doesn't exist any more, install closest 1011 to workaround
+    if [[ ! -d /usr/src/linux-azure-headers-4.15.0-1002 ]]; then
+        echo "installing linux-headers-fips..."
+        apt_get_install 5 10 120 linux-headers-fips || exit $LINUX_HEADER_INSTALL_TIMEOUT
+        ln -s /usr/src/linux-fips-headers-4.15.0-1011 /usr/src/linux-azure-headers-4.15.0-1002
+    fi
+
+    # now the fips packages/kernel are installed, clean up apt settings in the vhd,
+    # the VMs created on customers' subscriptions don't have access to UA repo
+    echo "detaching ua..."
+    retrycmd_if_failure 5 10 120 echo y | ua detach || $ERR_UA_DETACH
+
+    echo "removing ua tools..."
+    apt_get_purge 5 10 120 ubuntu-advantage-tools
+    rm -f /etc/apt/trusted.gpg.d/ua-client_ubuntu_stable.gpg
+    rm -f /etc/apt/trusted.gpg.d/ubuntu-advantage-esm-apps.gpg
+    rm -f /etc/apt/trusted.gpg.d/ubuntu-advantage-esm-infra-trusty.gpg
+    rm -f /etc/apt/trusted.gpg.d/ubuntu-advantage-fips.gpg
+    rm -f /etc/apt/sources.list.d/ua-client-ubuntu-stable-bionic.list
+    rm -f /etc/apt/sources.list.d/ubuntu-esm-apps.list
+    rm -f /etc/apt/sources.list.d/ubuntu-esm-infra.list
+    rm -f /etc/apt/sources.list.d/ubuntu-fips.list
+    rm -f /etc/apt/auth.conf.d/90ubuntu-advantage
+    apt_get_update || exit $ERR_APT_UPDATE_TIMEOUT
+
+    resolvconf=$(readlink -f /etc/resolv.conf)
```

Comment:
this is done in disable1804SystemdResolved for all 1804

Reply:
As we sync offline, disable1804SystemdResolved() in CSE is controlled by a toggle "disable-1804-systemd-resolved". But this toggle is not not enabled everywhere (enabled in INT, staging, and some customer subscriptions with CRIs). So the FIPS cluster created still has bad resolv.conf, and some pods are in crashloop because of this. I will keep these lines until toggle "disable-1804-systemd-resolved" is enabled everywhere.

Reply:
Could you make a comment here so we remember to remove?

Reply:
added a comment there.